lib/Twig/Environment.php

```diff
@@ -107,6 +108,8 @@ public function __construct(Twig_LoaderInterface $loader = null, $options = arra
             'optimizations' => -1,
         ), $options);
 
+        // cache can be object so remove it from options hash
+        $this->optionsHash = serialize(array_diff_key($options, array('cache' => null)));
```

Comment:
this implementation is wrong. All the settings impacting the compilation are mutable, so the hash should be recomputed on mutation